# 第7次作业
## 致理-生21 孙铭一 2022012361
#  Part III. 2.1 RNA-seq - Expression Matrix
## 1) RNA-seq中归一化基因表达值的几种基本计算方法

RNA-seq测序获得的原始read counts需要归一化，以消除测序深度和基因长度带来的偏差。常见的归一化方法包括：

- **CPM (Counts Per Million)**：
  - 公式：每个基因的raw count除以样本的总reads数再乘以一百万。
  - 适用于只考虑测序深度差异时的归一化。
  
- **TPM (Transcripts Per Million)**：
  - 公式：首先将每个基因的read counts除以基因长度（得到每kb的reads数），然后再归一化为每百万。
  - 更适合比较不同基因在同一样本中的表达量。

- **FPKM/RPKM (Fragments/Reads Per Kilobase of transcript per Million mapped reads)**：
  - 适用于成对reads测序 (FPKM) 或单端测序 (RPKM)。
  - 与TPM不同，归一化顺序为先计算总reads数，再考虑基因长度。

- **TMM (Trimmed Mean of M-values)**：
  - 由edgeR提出，修正样本中存在大量差异表达基因时导致的系统性偏差。

- **DESeq2 size factor normalization**：
  - 通过中位数归一化的方法，适用于差异表达分析。

---
## 2) 根据下图描述，Gene G的实际raw counts

根据mapping图：

- **Standard illumina (非链特异性)**：所有reads无论方向均计算，因此raw count = 13。
- **Ligation method (链特异性，reads2为sense strand)**：需要筛选sense方向上的reads，图中符合的为6条。
- **dUTPs method (链特异性，reads1为sense strand)**：需要筛选sense方向的reads，符合条件的是7条。

### 答案：
- Standard illumina: **E.13**
- Ligation method: **B.6**
- dUTPs method: **C.7**

---
## 3) 通过软件计算，判断给出文件shape02数据是来自哪一种sequencing protocols （strand nonspecific, strand specific - forward, strand specific - reverse)，并选择合适的参数计算shape02的read count matrix，给出AT1G09530基因(PIF3基因)上的counts数目。
```bash
PS C:\Users\86130> docker exec -it bioinfo_featurecount /bin/bash
root@featurecount_docker:/# cd /home/test
root@featurecount_docker:/home/test# ls
GTF  bam  result  share
root@featurecount_docker:/home/test# /usr/local/bin/infer_experiment.py -r GTF/Arabidopsis_thaliana.TAIR10.34.bed -i bam/Shape02.bam
Reading reference gene model GTF/Arabidopsis_thaliana.TAIR10.34.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled

This is PairEnd Data
Fraction of reads failed to determine: 0.0315
Fraction of reads explained by "1++,1--,2+-,2-+": 0.4769
Fraction of reads explained by "1+-,1-+,2++,2--": 0.4916
```
分析：
两个方向的reads比例接近50% / 50%。
说明reads在正义链和反义链上分布均匀。
结论：
Shape02数据是来自 **strand nonspecific**（非链特异性）测序协议。
### 选择合适参数计算Shape02的read count matrix
```bash
root@featurecount_docker:/home/test# /home/software/subread-2.0.3-source/bin/featureCounts -T 4 -s 0 -p -a GTF/Arabidopsis_thaliana.TAIR10.34.gtf -o Shape02_counts.txt bam/Shape02.bam

        ==========     _____ _    _ ____  _____  ______          _____
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
          v2.0.3

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 1 BAM file                                       ||
||                                                                            ||
||                           Shape02.bam                                      ||
||                                                                            ||
||             Output file : Shape02_counts.txt                               ||
||                 Summary : Shape02_counts.txt.summary                       ||
||              Paired-end : yes                                              ||
||        Count read pairs : no                                               ||
||              Annotation : Arabidopsis_thaliana.TAIR10.34.gtf (GTF)         ||
||      Dir for temp files : ./                                               ||
||                                                                            ||
||                 Threads : 4                                                ||
||                   Level : meta-feature level                               ||
||      Multimapping reads : not counted                                      ||
|| Multi-overlapping reads : not counted                                      ||
||   Min overlapping bases : 1                                                ||
||                                                                            ||
\\============================================================================//

//================================= Running ==================================\\
||                                                                            ||
|| Load annotation file Arabidopsis_thaliana.TAIR10.34.gtf ...                ||
||    Features : 313952                                                       ||
||    Meta-features : 32833                                                   ||
||    Chromosomes/contigs : 7                                                 ||
||                                                                            ||
|| Process BAM file Shape02.bam...                                            ||
||    Paired-end reads are included.                                          ||
||    The reads are assigned on the single-end mode.                          ||
||    Total alignments : 2730443                                              ||
||    Successfully assigned alignments : 2559170 (93.7%)                      ||
||    Running time : 0.03 minutes                                             ||
||                                                                            ||
|| Write the final count table.                                               ||
|| Write the read assignment summary.                                         ||
||                                                                            ||
|| Summary of counting results can be found in file "Shape02_counts.txt.summ  ||
|| ary"                                                                       ||
||                                                                            ||
\\============================================================================//

root@featurecount_docker:/home/test# ls
GTF  Shape02_counts.txt  Shape02_counts.txt.summary  bam  result  share
```

### 提取AT1G09530基因（PIF3基因）的counts数目
```bash
root@featurecount_docker:/home/test# grep "AT1G09530" Shape02_counts.txt
AT1G09530       1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1   3075768;3075768;3075768;3076401;3076401;3076401;3076459;3076459;3076459;3077173;3077173;3077173;3077173;3077378;3077378;3077378;3077378;3077378;3077378;3078346;3078346;3078346;3078346;3078346;3078346;3078545;3078545;3078545;3078545;3078545;3078545;3078843;3078843;3078843;3078843;3078843;3078843;3078984;3078984;3078984;3078984;3078984;3078984 3075852;3075852;3075852;3077286;3076808;3076748;3076808;3077286;3076748;3077286;3077286;3077286;3077286;3078257;3078257;3078257;3078257;3078257;3078257;3078453;3078453;3078453;3078453;3078453;3078453;3078610;3078610;3078610;3078610;3078610;3078610;3078908;3078908;3078908;3078908;3078908;3078908;3079544;3079544;3079544;3079654;3079654;3079654 +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+   2762    86
```
### PIF3基因（AT1G09530）counts数目为：86
---
## 4) tumor-transcriptome-demo.tar.gz提供了结肠癌(COAD)，直肠癌(READ)和食道癌(ESCA)三种癌症各50个样本的bam文件用featureCount计算产生的结果。请大家编写脚本将这些文件中的counts合并到一个矩阵中(行为基因，列为样本), 计算logCPM的Z-score，并用 heatmap 展示，提供代码和heatmap。根据heatmap可视化的结果，你认为这三种癌症中哪两种癌症的转录组是最相似的?
```bash
root@featurecount_docker:/home/test# ls
GTF  Shape02_counts.txt  Shape02_counts.txt.summary  bam  result  share  tumor-transcriptome-demo.tar.gz
root@featurecount_docker:/home/test# tar xvzf tumor-transcriptome-demo.tar.gz
root@featurecount_docker:/home/test# ls
GTF                 Shape02_counts.txt.summary  result  tumor-transcriptome-demo
Shape02_counts.txt  bam                         share   tumor-transcriptome-demo.tar.gz
PS C:\Users\86130> docker cp bioinfo_featurecount:/home/test/tumor-transcriptome-demo D:\大三下\生物信息学\PART3
Successfully copied 180MB to D:\大三下\生物信息学\PART3
```
```r
# 加载必要的包
library(edgeR)
library(pheatmap)

#设置工作目录
setwd("D:/大三下/生物信息学/PART3/tumor-transcriptome-demo")

#读取所有counts文件并合并成一个矩阵
count_files <- list.files(path = ".", pattern = "*.txt", recursive = TRUE, full.names = TRUE)

# 读取第一个文件建立基底
first_file <- read.table(count_files[1], header = TRUE, sep = "\t", comment.char = "#", row.names = 1)
all_counts <- data.frame(first_file[, ncol(first_file)])  # 取最后一列
colnames(all_counts) <- basename(gsub(".txt", "", count_files[1]))

# 循环读剩下的文件
for (i in 2:length(count_files)) {
  temp <- read.table(count_files[i], header = TRUE, sep = "\t", comment.char = "#", row.names = 1)
  temp_count <- temp[, ncol(temp)]
  all_counts <- cbind(all_counts, temp_count)
  colnames(all_counts)[i] <- basename(gsub(".txt", "", count_files[i]))
}

# 确认矩阵大小
dim(all_counts)

# 转为矩阵
counts_matrix <- as.matrix(all_counts)

# 计算CPM和log10CPM矩阵

CPM.matrix <- t(1000000 * t(counts_matrix) / colSums(counts_matrix))
log10.CPM.matrix <- log10(CPM.matrix + 1)

#计算Z-score矩阵并处理极端值
# 过滤掉标准差为0的基因（避免NaN问题）
nonzero_sd <- apply(log10.CPM.matrix, 1, sd) != 0
filtered_log10_CPM <- log10.CPM.matrix[nonzero_sd, ]

# 计算Z-score
z.scores <- (filtered_log10_CPM - rowMeans(filtered_log10_CPM)) / apply(filtered_log10_CPM, 1, sd)

# 将z-score裁剪到[-2, 2]，防止极端值影响展示
z.scores[z.scores > 2] <- 2
z.scores[z.scores < -2] <- -2

# 绘制Heatmap

pheatmap(z.scores,
         scale = "none",
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         show_colnames = FALSE,
         show_rownames = FALSE)
```
![image](https://github.com/user-attachments/assets/7b2c6be0-c6dd-4039-8fcf-9b0ac4ccfc98)

根据绘制出的heatmap结果，可以观察到：
- **COAD（结肠癌）** 和 **READ（直肠癌）** 的样本在列聚类（column clustering）上距离非常近，形成了一个明显的共同分支。
- **ESCA（食道癌）** 的样本则聚在另一边，与COAD和READ明显分离。

因此，可以得出结论：
> **COAD和READ这两种癌症的转录组最为相似。**
---

# Part III. 2.3 RNA-seq - Diff. Exp.





